core/libp2p/connection/yamux/yamux_stream.cpp

```diff
@@ -0,0 +1,303 @@
+          });
+    }
+
+    // each time a window size gets updated, that lambda will be called, and if
+    // that time the window is wide enough, finally write to the connection
+    return yamuxed_connection_->streamAddWindowUpdateNotifyee(
+        stream_id_,
+        [self{shared_from_this()}, in, bytes, some,
+         cb = std::move(cb)]() mutable {
+          if (self->send_window_size_ - bytes < 0) {
+            // not yet
+            return false;
+          }
+          self->yamuxed_connection_->streamWrite(
+              self->stream_id_, in, bytes, some,
+              [self, cb = std::move(cb)](auto &&res) {
+                self->is_writing_ = false;
+                if (res) {
+                  self->send_window_size_ -= res.value();
+                }
+                return cb(std::forward<decltype(res)>(res));
+              });
+          return true;
+        });
+  }
+
+  bool YamuxStream::isClosed() const noexcept {
+    return !is_readable_ && !is_writable_;
+  }
+
+  void YamuxStream::close(CloseCallbackFunc cb) {
+    if (is_writing_) {
+      return cb(Error::IS_WRITING);
+    }
+
+    is_writing_ = true;
+    return yamuxed_connection_->streamClose(
+        stream_id_, [self{shared_from_this()}, cb = std::move(cb)](auto &&res) {
+          self->is_writing_ = false;
+          cb(std::forward<decltype(res)>(res));
+        });
+  }
+
+  bool YamuxStream::isClosedForRead() const noexcept {
+    return !is_readable_;
+  }
+
+  bool YamuxStream::isClosedForWrite() const noexcept {
+    return !is_writable_;
+  }
+
+  void YamuxStream::reset(std::function<void(outcome::result<void>)> cb) {
+    if (is_writing_) {
+      return cb(Error::IS_WRITING);
+    }
+
+    is_writing_ = true;
+    return yamuxed_connection_->streamReset(
+        stream_id_, [self{shared_from_this()}, cb = std::move(cb)](auto &&res) {
+          self->is_writing_ = false;
+          self->resetStream();
+          cb(std::forward<decltype(res)>(res));
+        });
+  }
+
+  void YamuxStream::adjustWindowSize(
+      uint32_t new_size, std::function<void(outcome::result<void>)> cb) {
+    if (is_writing_) {
+      return cb(Error::IS_WRITING);
+    }
+
+    if (new_size > maximum_window_size_ || new_size < read_buffer_.size()) {
+      return cb(Error::INVALID_WINDOW_SIZE);
+    }
+
+    is_writing_ = true;
+    return yamuxed_connection_->streamAckBytes(
+        stream_id_, new_size - receive_window_size_,
+        [self{shared_from_this()}, cb = std::move(cb), new_size](auto &&res) {
+          self->is_writing_ = false;
+          if (!res) {
+            return cb(res.error());
+          }
+          self->receive_window_size_ = new_size;
+          cb(outcome::success());
+        });
+  }
+
+  void YamuxStream::resetStream() {
+    is_readable_ = false;
+    is_writable_ = false;
+  }
+
+  outcome::result<void> YamuxStream::commitData(gsl::span<const uint8_t> data,
+                                                size_t data_size) {
```

Comment:
why both data and data_size? isn't just span enough?

Reply:
Both parameters are necessary, as a buffer of MaximumWindowsize is preallocated in the YamuxedConnection, and span to it is passed to this method along with an exact number of bytes, which were read during the last read